09_simulations/models.py

```diff
@@ -0,0 +1,187 @@
+import numpy as np
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import (
+    roc_auc_score,
+    average_precision_score,
+    accuracy_score,
+    balanced_accuracy_score
+)
+from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
+from sklearn.neural_network import MLPClassifier
+from sklearn.pipeline import Pipeline
+
+def train_ridge(X_train,
+                y_train,
+                seed,
+                c_values=None,
+                n_folds=3,
+                max_iter=1000):
+
+    if c_values is None:
+        clf_parameters = {
+            "classify__C": [1e-6, 1e-5, 1e-4, 0.001, 0.01, 0.1, 1, 10, 100, 1000]
+        }
+    else:
+        clf_parameters = {
+            "classify__C": c_values
+        }
+
+    estimator = Pipeline(
+        steps=[
+            (
+                "classify",
+                LogisticRegression(
+                    random_state=seed,
+                    class_weight='balanced',
+                    penalty='l2',
+                    solver='lbfgs',
+                    max_iter=max_iter,
+                    tol=1e-3,
+                ),
+            )
+        ]
+    )
+    cv_pipeline = GridSearchCV(
+        estimator=estimator,
+        param_grid=clf_parameters,
+        n_jobs=-1,
+        cv=n_folds,
+        scoring='roc_auc',
+        return_train_score=True,
+    )
+
+    # Fit the model
+    cv_pipeline.fit(X=X_train, y=y_train)
+
+    return cv_pipeline
+
+
+def train_rf(X_train,
+             y_train,
+             seed,
+             n_folds=3,
+             max_iter=1000):
+
+    clf_parameters = {
+        "classify__max_depth": [3, 5, 8],
+        "classify__n_estimators": [10, 100, 200],
+    } 
+
+    estimator = Pipeline(
+        steps=[
+            (
+                "classify",
+                RandomForestClassifier(
+                    random_state=seed,
+                    class_weight='balanced',
+                ),
+            )
+        ]
+    )
+    cv_pipeline = GridSearchCV(
+        estimator=estimator,
+        param_grid=clf_parameters,
+        n_jobs=-1,
+        cv=n_folds,
+        scoring='roc_auc',
+        return_train_score=True,
+    )
+
+    # Fit the model
+    cv_pipeline.fit(X=X_train, y=y_train)
+
+    return cv_pipeline
+
+
+def train_mlp(X_train,
+              y_train,
+              params,
+              seed,
+              batch_size=50,
+              n_folds=3,
+              max_iter=1000,
+              search_n_iter=20):
+
+    import torch.optim
```

Comment:
Is there a reason you're importing them here?

Reply:
I've generally done this with my neural network code because getting pytorch installed and configured can be a bit annoying. If I put the import at the top, then anyone who runs the code will need to have pytorch installed, but if I put it in this function it will only be necessary if the user actually wants to fit a NN.